pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/RecoverableMultithreadedProcessor.java

```diff
@@ -0,0 +1,279 @@
+ * and writing each output event to another Pravega stream.
+ */
+public class RecoverableMultithreadedProcessor {
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(RecoverableMultithreadedProcessor.class);
+
+    public static class Master implements Runnable {
+        private static final org.slf4j.Logger log = LoggerFactory.getLogger(Master.class);
+
+        private static final String PRAVEGA_CHECKPOINT_FILE_NAME = "pravega-checkpoint";
+        private static final String LATEST_CHECKPOINT_NAME_FILE_NAME = "latest";
+
+        private final String scope;
+        private final String inputStreamName;
+        private final String outputStreamName;
+        private final URI controllerURI;
+        private final int numWorkers;
+        private final String readerGroupName;
+        private final ReaderGroup readerGroup;
+        private final ScheduledExecutorService initiateCheckpointExecutor;
+        private final ScheduledExecutorService performCheckpointExecutor;
+        private final ExecutorService workerExecutor;
+        private final ReaderGroupManager readerGroupManager;
+        private final Path checkpointRootPath = Parameters.getCheckpointRootPath();
+        private final Path latestCheckpointNamePath = checkpointRootPath.resolve(LATEST_CHECKPOINT_NAME_FILE_NAME);
+        private final long checkpointPeriodMs = Parameters.getCheckpointPeriodMs();
+        private final long checkpointTimeoutMs = Parameters.getCheckpointTimeoutMs();
+
+        public Master(String scope, String inputStreamName, String outputStreamName, URI controllerURI, int numWorkers) throws Exception {
+            this.scope = scope;
+            this.inputStreamName = inputStreamName;
+            this.outputStreamName = outputStreamName;
+            this.controllerURI = controllerURI;
+            this.numWorkers = numWorkers;
+
+            // Create streams.
+            try (StreamManager streamManager = StreamManager.create(controllerURI)) {
+                streamManager.createScope(scope);
+                StreamConfiguration streamConfig = StreamConfiguration.builder()
+                        .scalingPolicy(ScalingPolicy.byEventRate(
+                                Parameters.getTargetRateEventsPerSec(),
+                                Parameters.getScaleFactor(),
+                                Parameters.getMinNumSegments()))
+                        .build();
+                streamManager.createStream(scope, inputStreamName, streamConfig);
+                // Since we start reading the input stream from the earliest event, we must delete the output stream.
+                streamManager.sealStream(scope, outputStreamName);
+                streamManager.deleteStream(scope, outputStreamName);
+                streamManager.createStream(scope, outputStreamName, streamConfig);
+            }
+
+            // Create a reader group manager. It must remain open to allow manual checkpoints to work.
```

Comment:
This comment is not accurate. Once a checkpoint is initiated all readers will get it.

Reply:
If I close the ReaderGroupManager immediately after getReaderGroup, calls to readerGroup.getOnlineReaders and readerGroup.initiateCheckpoint will never return. Is this a bug?

Reply:
No, it's not a bug. But it doesn't make the comment accurate. You only need to keep the reader group manager open to find out when the checkpoint has completed, not for it to actually complete.

Reply:
The comment is now "It must remain open so that readerGroup.initiateCheckpoint returns".